zaza/openstack/charm_tests/keystone/setup.py

```diff
@@ -192,17 +192,34 @@ def wait_for_url(url, ok_codes=None):
             assert r.status_code in ok_codes
 
 
+def wait_for_client():
```

Comment:
nitpick: the function's name is a bit strange ... it's waiting, but also returning a value ... the "waiting" seems to be irrelevant from the caller's perspective, why not call it "get_keystone_client"?, or maybe re-use get_keystone_client()[0] and wrap that one with a retry() decorator?

[0] https://github.com/openstack-charmers/zaza-openstack-tests/blob/master/zaza/openstack/utilities/openstack.py#L616

Reply:
I agree that the function's name is wrong.  I don't think that `get_keystone_client()` can be used as it seems more basic that this function; i.e. it's using overcloud auth rather than openrc.

I think this function ought to be in `z.o.utilities.openstack.py` though along with the other keystone client functions which would also open a space for refactoring at a future point.

Finally, the  `wait_for_url()` function is also really generic and probably could also live in a utility module; the `wait_for_all_endpoints()` a bit more arguable as to where it should go.  It's specific enother to be testing keystone ... but also generic enough that other tests might also want to wait on the endpoints being available as a utility.